File: lab2/binary.cpp
#define CATCH_CONFIG_MAIN
#include "catch.hpp"
#include <fstream>
#include <vector>
using namespace std;
int binarySearch(int arr[], int low, int high, int x) {
  if (high >= low) {
    int mid = low + (high - low) / 2;

    if (arr[mid] == x)
      return mid;

    if (arr[mid] > x)
      return binarySearch(arr, low, mid - 1, x);

    return binarySearch(arr, mid + 1, high, x);
  }
  return -1;
}

vector<int> readData(const string &filename, int &query) {
  ifstream infile(filename);
  if (!infile) {
    throw runtime_error("Unable to open file");
  }

  int n;
  infile >> n;
  vector<int> arr(n);
  for (int i = 0; i < n; ++i) {
    infile >> arr[i];
  }

  infile >> query;
  infile.close();
  return arr;
}

TEST_CASE("Test", "[binarySearch]") {
  SECTION("test1") {
    int query;
    vector<int> arr = readData("../binary.txt", query);
    REQUIRE(binarySearch(arr.data(), 0, arr.size() - 1, query) == 3);
  }
  SECTION("test2") {
    int query;
    vector<int> arr = readData("../binary1.txt", query);
    REQUIRE(binarySearch(arr.data(), 0, arr.size() - 1, query) == -1);
  }
}

File: lab2/binary.py
import unittest


def binarySearch(arr, low, high, x):

    if high >= low:

        mid = low + (high - low) // 2

        if arr[mid] == x:
            return mid

        elif arr[mid] > x:
            return binarySearch(arr, low, mid - 1, x)

        else:
            return binarySearch(arr, mid + 1, high, x)

    else:
        return -1


def read_data(filename):
    with open(filename, "r") as file:
        n = int(file.readline().strip())
        arr = list(map(int, file.readline().strip().split()))
        query = int(file.readline().strip())
    return arr, query


class Test(unittest.TestCase):

    def test1(self):
        arr, query = read_data("../binary.txt")
        result = binarySearch(arr, 0, len(arr) - 1, query)
        self.assertEqual(result, 3)

    def test2(self):
        arr, query = read_data("../binary1.txt")
        result = binarySearch(arr, 0, len(arr) - 1, query)
        self.assertEqual(result, -1)


if __name__ == "__main__":
    unittest.main()
